pkg/auth/authenticator/password/ldappassword/ldap.go

```diff
@@ -0,0 +1,240 @@
+	AttributeID []string
+}
+
+// Authenticator validates username/passwords against an LDAP v3 server
+type Authenticator struct {
+	providerName string
+	options      Options
+	mapper       authapi.UserIdentityMapper
+}
+
+// New returns an authenticator which will validate usernames/passwords using LDAP.
+func New(providerName string, options Options, mapper authapi.UserIdentityMapper) (authenticator.Password, error) {
+	auth := &Authenticator{
+		providerName: providerName,
+		options:      options,
+		mapper:       mapper,
+	}
+	return auth, nil
+}
+
+// AuthenticatePassword validates the given username and password against an LDAP server
+func (a *Authenticator) AuthenticatePassword(username, password string) (user.Info, bool, error) {
+	identity, ok, err := a.getIdentity(username, password)
+	if err != nil {
+		return nil, false, err
+	}
+	if !ok {
+		return nil, false, nil
+	}
+
+	user, err := a.mapper.UserFor(identity)
+	glog.V(4).Infof("Got userIdentityMapping: %#v", user)
+	if err != nil {
+		return nil, false, fmt.Errorf("Error creating or updating mapping for: %#v due to %v", identity, err)
+	}
+
+	return user, true, nil
+
+}
+
+// connect returns an established ldap connection, or an error if the connection could not be made (or successfully upgraded to TLS)
+// if no error is returned, the caller is responsible for closing the connection
+func (a *Authenticator) connect() (*ldap.Conn, error) {
+	switch a.options.URL.Scheme {
+	case ProtocolLDAP:
+		l, err := ldap.Dial("tcp", a.options.URL.Host)
+		if err != nil {
+			return nil, err
+		}
+
+		// If an insecure connection is desired, we're done
+		if a.options.Insecure {
+			return l, nil
+		}
+
+		// Attempt to upgrade to TLS
+		if err := l.StartTLS(a.options.TLSConfig); err != nil {
+			// We're returning an error on a successfully opened connection
+			// We are responsible for closing the open connection
+			l.Close()
+			return nil, err
+		}
+
+		return l, nil
+
+	case ProtocolLDAPS:
+		return ldap.DialTLS("tcp", a.options.URL.Host, a.options.TLSConfig)
+
+	default:
+		return nil, fmt.Errorf("unsupported scheme %q", a.options.URL.Scheme)
+	}
+}
+
+// getIdentity looks up a username in an LDAP server, and attempts to bind to the user's DN using the provided password
+func (a *Authenticator) getIdentity(username, password string) (authapi.UserIdentityInfo, bool, error) {
+	if len(username) == 0 || len(password) == 0 {
+		return nil, false, nil
+	}
+
+	// Make the connection
+	l, err := a.connect()
+	if err != nil {
+		return nil, false, err
+	}
+	defer l.Close()
+
+	// If specified, bind the username/password for search phase
+	if len(a.options.BindDN) > 0 {
+		if err := l.Bind(a.options.BindDN, a.options.BindPassword); err != nil {
+			return nil, false, err
+		}
+	}
+
+	// & together the filter specified in the LDAP options with the user-specific filter
+	filter := fmt.Sprintf("(&%s(%s=%s))",
```

Comment:
Not that I think any sane person would do it, but do we want to support arbitrary filters that allow for multiple ands/ors/nots, etc?  Things like, "I want to use this ldap, but only people belonging to this department have access and I can control that with a clever filter in my authenticator."